avro-fastserde/src/main/java/com/linkedin/avro/fastserde/Utils.java

```diff
@@ -26,6 +28,16 @@
   // Cache the mapping between Schema and the corresponding fingerprint
   private static final Map<Schema, Long> SCHEMA_IDS_CACHE = new ConcurrentHashMap<>();
 
+  // Schema#hashCode is not performance efficient in Avro 1.4. It has been improved
+  // by caching hashcode in the later version. Change to use IdentityHashMap for
+  // Avro 1.4 helps to improve the cache lookup performance. However, it requires
+  // the client to resuse the schame instance. Or, it will ventually lead to
+  // "Too many schema objects" issue
+  // TODO - LRU Cache with size limit
+  private static final Map<Schema, Long> SCHEMA_IDS_CACHE_AVRO_14 = new IdentityHashMap<>();
```

Comment:
as you stated, this has a memory issue. you could use WeakIdentityHashMap for this. while its not a jdk class there are implementations in various places you could port over - https://docs.jboss.org/jbossas/javadoc/4.0.4/common/org/jboss/util/collection/WeakIdentityHashMap.html https://www.gnu.org/software/kawa/api/gnu/kawa/util/WeakIdentityHashMap.html even avro 1.8 has one - https://avro.apache.org/docs/1.8.0/api/java/org/apache/avro/util/WeakIdentityHashMap.html (although they warn its non standard)

Reply:
yet another potential implementation - https://github.com/gfx/Java-WeakIdentityHashMap

Reply:
Good idea. Done.